fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java

```diff
@@ -0,0 +1,268 @@
+    }
+
+    /**
+     * @param node
+     * @return the id to the Patient resource referenced by this node (assuming it is a Reference with a valid
+     *         reference value); otherwise null
+     */
+    private String getPatientRefVal(FHIRPathNode node) {
+        if (!node.isElementNode() || !node.asElementNode().element().is(Reference.class)) {
+            throw new IllegalStateException("Patient compartment inclusionCriteria expression has returned a non-Reference");
+        }
+        Reference reference = node.asElementNode().element().as(Reference.class);
+        if (reference.getReference() != null && reference.getReference().hasValue()) {
+            String refVal = reference.getReference().getValue();
+            if (refVal != null && refVal.startsWith(PATIENT_REF_PREFIX)) {
+                return refVal.substring(PATIENT_REF_PREFIX.length());
+            }
+        } else if (log.isLoggable(Level.FINE)){
+            log.fine("Skipping non-patient / non-relative reference: '" + reference + "'");
+        }
+
+        return null;
+    }
+
+    /**
+     * @param permission
+     * @param requiredPermission
+     * @return true if the grantedPermission includes the requiredPermission; otherwise false
+     */
+    private boolean hasPermission(Permission grantedPermission, Permission requiredPermission) {
+        if (grantedPermission == Permission.ALL) {
+            return true;
+        } else {
+            return grantedPermission == requiredPermission;
+        }
+    }
+
+    private String getAccessToken() throws FHIRPersistenceInterceptorException {
+        List<String> list = FHIRRequestContext.get().getHttpHeaders().get("Authorization");
+        if (list.size() != 1) {
+            throw new FHIRPersistenceInterceptorException("Request must contain exactly one Authorization header.");
+        }
+        String header = list.get(0);
+
+        if (!header.startsWith(BEARER_TOKEN_PREFIX)) {
+            throw new FHIRPersistenceInterceptorException("Authorization header must carry a Bearer token");
+        }
+
+        return header.substring(BEARER_TOKEN_PREFIX.length()).trim();
+    }
+
+    private List<Scope> getScopesFromToken(DecodedJWT jwt) {
+        Claim claim = jwt.getClaim("scope");
+        if (claim.isNull()) {
+            throw new IllegalArgumentException("Authorization token is missing 'scope' claim");
+        }
+
+        List<String> scopeStrings;
+        String claimString = claim.asString();
+
+        if (claimString != null) {
+            scopeStrings = Arrays.asList(claim.asString().split("\\s+"));
+        } else {
+            log.fine("Found scope claim was expected to be a string but is not; processing as a list");
+            scopeStrings = claim.asList(String.class);
+        }
+
+        return scopeStrings.stream()
+                .filter(s -> s.matches(Scope.SCOPE_STRING_REGEX))
+                .map(s -> new Scope(s))
+                .collect(Collectors.toList());
```

Comment:
small set may be best to have a for loop.... same performance question

Reply:
same response.  can change if needed, but I think its fine.